### PR TITLE
Implement force_data_skipping_indices setting

### PR DESCRIPTION
**Pull request by azat**
Changelog category (leave one):
- Improvement

Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Implement force_data_skipping_indices setting

P.S. named it indices by analogy to allow_experimental_data_skipping_indices (obsolete already, removed in #15643)

<details>

HEAD:
- e00008e61dc13cedb8e9f4c8afd6501f9f7b6a6e

</details>

### CODE REVIEW

**Comment by alexey-milovidov** [comment]
If at least one index is required then better to name it `force_data_skipping_index`?

**Comment by alexey-milovidov** [comment]
So, it's for the list of indices to force...

File: docs/en/operations/settings/settings.md [hunk 1]
```diff
@@ -70,6 +70,12 @@ Works with tables in the MergeTree family.
 
 If `force_primary_key=1`, ClickHouse checks to see if the query has a primary key condition that can be used for restricting data ranges. If there is no suitable condition, it throws an exception. However, it does not check whether the condition reduces the amount of data to read. For more information about data ranges in MergeTree tables, see [MergeTree](../../engines/table-engines/mergetree-family/mergetree.md).
 
+## force\_data\_skipping\_indices {#settings-force_data_skipping_indices}
+
+Disables query execution if passed data skipping indices wasn't used.
```
Comment (alexey-milovidov) [inline docs/en/operations/settings/settings.md:75]: But it's unclear from docs.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 2]
```diff
@@ -552,6 +554,31 @@ Pipe MergeTreeDataSelectExecutor::readFromParts(
...
     }
 
+    if (settings.force_data_skipping_indices.changed)
+    {
+        std::unordered_set<std::string> useful_indices_names;
+        for (const auto & [ index_helper, _ ]: useful_indices)
+            useful_indices_names.insert(index_helper->index.name);
+
+        std::vector<std::string> forced_indices;
+        boost::split(forced_indices,
```
Comment (alexey-milovidov) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:564]: Maybe use full featured parser that will support whitespaces and quoted identifiers?

Comment (alexey-milovidov) [reply src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:564]: Example:
```
index1, `my index`
```

**Review by alexey-milovidov — COMMENTED** [review]
The idea LGTM, a few minor changes required.

**Comment by alexey-milovidov** [comment]
I've fixed "Arcadia".

**Comment by azat** [comment]
>I've fixed "Arcadia".

Will try a better fix (w/o preprocessor)

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp (hunk 2 above)
Comment (azat) [reply src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp]: Ok (although looks a bit over kill for this feature to me).

File: docs/en/operations/settings/settings.md (hunk 1 above)
Comment (azat) [reply docs/en/operations/settings/settings.md:75]: What exactly? I've added a few examples hope that they will make things cleaner...

**Comment by azat** [comment]
>If at least one index is required then better to name it force_data_skipping_index?

Nope, all of them should be in use (and the test shows this, also added a few examples to the documentation to underline this there too).

**Comment by azat** [comment]
>Yandex synchronization check (only for Yandex employees)

Hm it fails, although I removed structured binding, @alexey-milovidov can you please take a look?

**Comment by alexey-milovidov** [comment]
> Hm it fails, although I removed structured binding

It's about `contains` method.
Don't worry and just go ahead with the ifdef.

File: src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp [hunk 3]
```diff
@@ -552,6 +555,38 @@ Pipe MergeTreeDataSelectExecutor::readFromParts(
...
+        if (forced_indices.empty())
+            throw Exception(ErrorCodes::CANNOT_PARSE_TEXT, "No indices parsed from force_data_skipping_indices ('{}')", indices);
+
+        std::unordered_set<std::string> useful_indices_names;
+        for (const auto & useful_index : useful_indices)
+            useful_indices_names.insert(useful_index.first->index.name);
+
+        for (const auto & index_name : forced_indices)
+        {
+            if (!useful_indices_names.contains(index_name))
```
Comment (azat) [inline src/Storages/MergeTree/MergeTreeDataSelectExecutor.cpp:581]: replace with `count` for arcadia

**Review by alexey-milovidov — APPROVED** [review]

**Comment by xqliang** [comment]
Docs needs updated here:

https://clickhouse.com/docs/en/guides/improving-query-performance/skipping-indexes/#skip-index-settings

* **force_data_skipping_indexes** (comma separated list of index names). This setting can be used to prevent some kinds of inefficient queries. In circumstances where querying a table is too expensive unless a skip index is used, using this setting with one or more index names will return an exception for any query that does not use the listed index. This would prevent poorly written queries from consuming server resources.

**Pull request merged**